packages/react-async/src/component.tsx

```diff
@@ -1,135 +1,330 @@
   }
 
+  KeepFresh.displayName = `Async.KeepFresh(${displayName})`;
+
   // Once we upgrade past TS 3.1, this will no longer be necessary,
   // because you can statically assign values to functions and TS
   // will know to augment its type
   const FinalComponent: AsyncComponentType<
+    ComponentType<Props>,
     Props,
-    PreloadProps,
-    PrefetchProps,
-    KeepFreshProps
-  > = AsyncComponent as any;
+    PreloadOptions,
+    PrefetchOptions,
+    KeepFreshOptions
+  > = Async as any;
+
+  Reflect.defineProperty(FinalComponent, 'resolver', {
+    value: resolver,
+    writable: false,
+  });
+
+  Reflect.defineProperty(FinalComponent, 'Preload', {
+    value: Preload,
+    writable: false,
+  });
+
+  Reflect.defineProperty(FinalComponent, 'Prefetch', {
+    value: Prefetch,
+    writable: false,
+  });
 
-  FinalComponent.Preload = Preload;
-  FinalComponent.Prefetch = Prefetch;
-  FinalComponent.KeepFresh = KeepFresh;
+  Reflect.defineProperty(FinalComponent, 'KeepFresh', {
+    value: KeepFresh,
+    writable: false,
+  });
+
+  Reflect.defineProperty(FinalComponent, 'usePreload', {
+    value: usePreload,
+    writable: false,
+  });
+
+  Reflect.defineProperty(FinalComponent, 'usePrefetch', {
+    value: usePrefetch,
+    writable: false,
+  });
+
+  Reflect.defineProperty(FinalComponent, 'useKeepFresh', {
+    value: useKeepFresh,
+    writable: false,
+  });
 
   return FinalComponent;
 }
 
+function noopUse() {
+  return () => {};
+}
+
 function noopRender() {
   return null;
 }
 
-function splitProps<Props>(
-  props: Props & ConstantProps,
-): [Props, AsyncPropsRuntime] {
-  const {async, ...rest} = props as any;
-  return [rest, async];
+const DEFAULT_DISPLAY_NAME = 'Component';
+const FILENAME_REGEX = /([^/]*)\.\w+$/;
+
+function displayNameFromId(id?: string) {
+  if (!id) {
+    return DEFAULT_DISPLAY_NAME;
+  }
+
+  const match = id.match(FILENAME_REGEX);
+  return match ? match[1] : DEFAULT_DISPLAY_NAME;
+}
+
+function defaultRenderError(error: Error) {
+  if (error) {
+    throw error;
+  }
+
+  return null;
+}
+
+function Loader<T>({
+  defer,
+  load,
+  props,
+}: {
+  defer?: DeferTiming | ((props: T) => boolean);
+  load(): void;
+  props: T;
+}) {
+  const handleIntersection = useCallback(
+    ({isIntersecting = true}) => {
+      if (isIntersecting) {
+        load();
+      }
+    },
+    [load],
+  );
+
+  useEffect(
+    () => {
+      if (defer == null || defer === DeferTiming.Mount) {
+        load();
+      } else if (typeof defer === 'function' && defer(props)) {
+        load();
+      }
+    },
+    [defer, load, props],
+  );
+
+  if (typeof defer === 'function') {
+    return null;
+  }
+
+  switch (defer) {
+    case DeferTiming.Idle:
+      return <OnIdle perform={load} />;
+    case DeferTiming.InViewport:
+      return (
+        <IntersectionObserver
+          threshold={0}
+          onIntersectionChange={handleIntersection}
+        />
+      );
+    default:
+      return null;
```

Comment:
so if I understand correctly, if I pass `defer={"Some madeup value" as any}` this will fail silently / never load?

Could we throw an error here?

Reply:
Do we usually guard against people intentionally ignoring the type system for no reason?

Reply:
We also don't usually have a code path for it /shrug

Reply:
I guess so, I think if this were gone there would be the inverse question of "what happens if it doesn't match the provided cases?"